include/xsimd/types/xsimd_fallback.hpp

```diff
@@ -0,0 +1,1052 @@
+****************************************************************************/
+
+#ifndef XSIMD_FALLBACK_HPP
+#define XSIMD_FALLBACK_HPP
+
+#include <algorithm>
+#include <array>
+#include <cmath>
+#include <utility>
+
+#include "xsimd_base.hpp"
+
+namespace xsimd
+{
+
+    /***********************************************************
+     * Generic fallback implementation of batch and batch_bool *
+     *                                                         *
+     * Basically, generate a scalar loop and cross fingers:    *
+     * maybe the compiler will autovectorize, maybe not.       *
+     ***********************************************************/
+
+    /**************************************
+     * Ugly shared implementation details *
+     **************************************/
+
+    // TODO: Some of these are pretty general and should probably go into xtl
+    namespace detail
+    {
+        // Boilerplate to get index_sequence in c++11
+        template <typename T>
+        struct identity { using type = T; };
+
+        #ifdef __cpp_lib_integer_sequence
+            using std::integer_sequence;
+            using std::index_sequence;
+            using std::make_index_sequence;
+            using std::index_sequence_for;
+        #else
+            template <typename T, T... Is>
+            struct integer_sequence {
+            using value_type = T;
+            static constexpr std::size_t size() noexcept { return sizeof...(Is); }
+            };
+
+            template <std::size_t... Is>
+            using index_sequence = integer_sequence<std::size_t, Is...>;
+
+            template <typename Lhs, typename Rhs>
+            struct make_index_sequence_concat;
+
+            template <std::size_t... Lhs, std::size_t... Rhs>
+            struct make_index_sequence_concat<index_sequence<Lhs...>,
+                                            index_sequence<Rhs...>>
+              : identity<index_sequence<Lhs..., (sizeof...(Lhs) + Rhs)...>> {};
+
+            template <std::size_t N>
+            struct make_index_sequence_impl;
+
+            template <std::size_t N>
+            using make_index_sequence = typename make_index_sequence_impl<N>::type;
+
+            template <std::size_t N>
+            struct make_index_sequence_impl
+              : make_index_sequence_concat<make_index_sequence<N / 2>,
+                                           make_index_sequence<N - (N / 2)>> {};
+
+            template <>
+            struct make_index_sequence_impl<0> : identity<index_sequence<>> {};
+
+            template <>
+            struct make_index_sequence_impl<1> : identity<index_sequence<0>> {};
+
+            template <typename... Ts>
+            using index_sequence_for = make_index_sequence<sizeof...(Ts)>;
+        #endif
+
+        // Tools for reinterpreting stuff as an unsigned integer
+        template <typename T>
+        struct as_unsigned;
```

Comment:
Some of these structures are already defined in `xsimd/types/xsimd_utils.hpp` (see `as_integer` and `as_unsigned_integer`), so it would be better to move some of these definitions should be moved there (and the names should be unified).

Reply:
Done